src/main/java/org/apache/datasketches/kll/KllHelper.java

```diff
@@ -19,32 +19,192 @@
+   * @param n the desired stream length
+   * @param sketchType the given sketch type (DOUBLES_SKETCH or FLOATS_SKETCH)
+   * @param printGrowthScheme if true the entire growth scheme of the sketch will be printed.
+   * @return GrowthStats with the final values of the growth scheme
    */
-  static int ubOnNumLevels(final long n) {
-    return 1 + Long.numberOfTrailingZeros(floorPowerOf2(n));
+  public static GrowthStats getGrowthSchemeForGivenN(
+      final int k,
+      final int m,
+      final long n,
+      final SketchType sketchType,
+      final boolean printGrowthScheme) {
+    int numLevels = 0;
+    LevelStats lvlStats;
+    final GrowthStats gStats = new GrowthStats();
+    gStats.k = k;
+    gStats.m = m;
+    gStats.givenN = n;
+    gStats.sketchType = sketchType;
+    if (printGrowthScheme) {
+      println("GROWTH SCHEME:");
+      println("Given SketchType: " + sketchType.toString());
+      println("Given K         : " + k);
+      println("Given M         : " + m);
+      println("Given N         : " + n);
+      printf("%10s %10s %20s %13s %15s\n", "NumLevels", "MaxItems", "MaxN", "CompactBytes", "UpdatableBytes");
+    }
+    int compactBytes;
+    int updatableBytes;
+    do {
+      numLevels++;
+      lvlStats = getFinalSketchStatsAtNumLevels(k, m, numLevels, false);
+      final int maxItems = lvlStats.items;
+      final long maxN = lvlStats.n;
+      if (sketchType == DOUBLES_SKETCH) {
+        compactBytes = maxItems * Double.BYTES + numLevels * Integer.BYTES + 2 * Double.BYTES + DATA_START_ADR_DOUBLE;
+        updatableBytes = compactBytes + Integer.BYTES;
+      } else {
+        compactBytes = maxItems * Float.BYTES + numLevels * Integer.BYTES + 2 * Float.BYTES + DATA_START_ADR_FLOAT;
+        updatableBytes = compactBytes + Integer.BYTES;
+      }
+      if (printGrowthScheme) {
+        printf("%10d %,10d %,20d %,13d %,15d\n", numLevels, maxItems, maxN, compactBytes, updatableBytes);
+      }
+    } while (lvlStats.n < n);
+    gStats.maxN = lvlStats.n;
+    gStats.numLevels = lvlStats.numLevels;
+    gStats.maxItems = lvlStats.items;
+    gStats.compactBytes = compactBytes;
+    gStats.updatableBytes = updatableBytes;
+    return gStats;
+  }
+
+  /**
+   * Given k, m, numLevels, this computes the item capacity of a single level.
+   * @param k the given user sketch configuration parameter
+   * @param m the given user sketch configuration parameter
+   * @param numLevels the given number of levels of the sketch
+   * @param level the specific level to compute its item capacity
+   * @return LevelStats with the computed N and items for the given level.
+   */
+  public static LevelStats getLevelCapacityItems(
+      final int k,
+      final int m,
+      final int numLevels,
+      final int level) {
+    final int items = KllHelper.levelCapacity(k, numLevels, level, m);
+    final long n = (long)items << level;
+    return new LevelStats(n, numLevels, items);
+  }
```

Comment:
Yes. To get a good understanding of how this works I would suggest the following.
- Start with a simple KllFloatsSketch(20) and fill it with 20 items. 
- Then print sketch.toString(true, true);  It will print the  structure of the sketch.
- Now add one more item and again print sketch.toString(true, true); 
This will show you how the sketch structure has changed.  It had to add a new level.

Keep doing this and you will see that each new level is about 2/3 the size of the previous level, 
until it reaches _m_, which is the minimum size of a level. All new levels after that will be the same
size _m_.  

This will also show you when the sketch compresses the current levels.

Reply:
Interesting - I plugged the level capacity code into a REPL and was playing around with the values but I'll do the same for your procedure above.